rust/crd/src/authentication.rs

```diff
@@ -91,64 +119,838 @@ pub fn default_user_registration_role() -> String {
+        let mut user_registration_role = None;
+        let mut sync_roles_at = None;
+
+        let mut auth_class_names = BTreeSet::new();
+
+        for entry in auth_details {
+            let auth_class_name = entry.common.authentication_class_name();
+
+            let is_new_auth_class = auth_class_names.insert(auth_class_name);
+            ensure!(
+                is_new_auth_class,
+                DuplicateAuthenticationClassReferencesNotAllowedSnafu { auth_class_name }
+            );
+
+            let auth_class = resolve_auth_class(entry.common.clone())
+                .await
+                .context(AuthenticationClassRetrievalFailedSnafu)?;
+
+            match &auth_class.spec.provider {
+                AuthenticationClassProvider::Ldap(provider) => {
+                    let resolved_auth_class = SupersetAuthenticationClassResolved::Ldap {
+                        provider: provider.to_owned(),
+                    };
 
-        // TODO: adapt if multiple authentication classes are supported by superset.
-        //    This is currently not possible due to the Flask App Builder not supporting it.
-        if self.authentication.len() > 1 {
-            return Err(Error::MultipleAuthenticationClassesProvided);
+                    if let Some(other) = resolved_auth_classes.first() {
+                        ensure!(
+                            mem::discriminant(other) == mem::discriminant(&resolved_auth_class),
+                            MultipleAuthenticationTypesNotSupportedSnafu
+                        );
+                    }
+
+                    ensure!(
+                        resolved_auth_classes.is_empty(),
+                        MultipleLdapProvidersNotSupportedSnafu
+                    );
+
+                    resolved_auth_classes.push(resolved_auth_class);
+                }
+                AuthenticationClassProvider::Oidc(provider) => {
+                    let resolved_auth_class =
+                        SupersetClientAuthenticationDetailsResolved::from_oidc(
+                            auth_class_name,
+                            provider,
+                            entry,
+                        )?;
+
+                    if let Some(other) = resolved_auth_classes.first() {
+                        ensure!(
+                            mem::discriminant(other) == mem::discriminant(&resolved_auth_class),
+                            MultipleAuthenticationTypesNotSupportedSnafu
+                        );
+                    }
+
+                    resolved_auth_classes.push(resolved_auth_class);
+                }
+                _ => {
+                    return Err(Error::AuthenticationProviderNotSupported {
+                        auth_class_name: auth_class_name.to_owned(),
+                        provider: auth_class.spec.provider.to_string(),
+                    });
+                }
+            }
+
+            match user_registration {
```

Comment:
This code ensures that `user_registration` is to the same value in all entries of the `auth_details`.

`user_registration` is initialized with `None`. Then the `auth_details` are iterated over. When the first entry of the `auth_details` is processed, then `user_registration` is still `None` and it is set to the value of the entry. When the following entries are processed, then `user_registration` contains the value of the first entry, which must be equal to the value of the processed entry, otherwise an error is thrown.

There are test cases for each of these error conditions, e.g. https://github.com/stackabletech/superset-operator/blob/71007aae441a8aae5bd8d8b10914a9d491fca31f/rust/crd/src/authentication.rs#L611-L654

(`user_registration` defaults to `true`.)

Reply:
Ah! The `for` loop! Of course.